addons/binding/org.openhab.binding.squeezebox/src/main/java/org/openhab/binding/squeezebox/discovery/SqueezeBoxServerDiscoveryService.java

```diff
@@ -0,0 +1,275 @@
+import org.eclipse.smarthome.core.thing.ThingTypeUID;
+import org.eclipse.smarthome.core.thing.ThingUID;
+import org.openhab.binding.squeezebox.SqueezeBoxBindingConstants;
+import org.openhab.binding.squeezebox.internal.utils.HttpUtils;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+/**
+ * SqueezeBoxServerDiscoveryService sends a UDP message on port 3483 to the
+ * network in order to trigger listening Squeeze Servers to advertise
+ * themselves.
+ * 
+ * @author Dan Cunningham
+ *
+ */
+public class SqueezeBoxServerDiscoveryService extends AbstractDiscoveryService {
+	private final Logger logger = LoggerFactory
+			.getLogger(SqueezeBoxServerDiscoveryService.class);
+
+	/**
+	 * Give us 5 seconds to find servers
+	 */
+	private static int TIMEOUT = 5;
+	/**
+	 * Port SqueezeServers listen on for discovery requestd
+	 */
+	private static final int DISCO_PORT = 3483;
+	/**
+	 * Discovery packet we send on the network to find Squeeze Servers
+	 */
+	private String discoPacket = "eIPAD\000NAME\000JSON\000VERS\000UUID\000JVID\006\012\034\056\078\012\034";
+	/**
+	 * Discovery replay packet containing the name, web port, version and UUID
+	 * of a Squeeze Server
+	 */
+	private Pattern replyPattern = Pattern
+			.compile("ENAME(.*?)JSON(.*?)VERS(.*?)UUID\\$(.*?)");
+	/**
+	 * When true we will keep sending out packets
+	 */
+	private boolean discoveryRunning;
+	/**
+	 * our Discovery socket
+	 */
+	private DatagramSocket socket;
+
+	/**
+	 * Looks for Squeeze Servers on the local network
+	 */
+	public SqueezeBoxServerDiscoveryService() {
+		super(Collections.singleton(SQUEEZEBOXSERVER_THING_TYPE), TIMEOUT, true);
+	}
+
+	@Override
+	public Set<ThingTypeUID> getSupportedThingTypes() {
+		return Collections.singleton(SQUEEZEBOXSERVER_THING_TYPE);
+	}
+
+	@Override
+	protected void startScan() {
+		scanServer();
+	}
+
+	@Override
+	protected void startBackgroundDiscovery() {
+		scanServer();
+	}
+
+	@Override
+	protected void stopScan() {
+		super.stopScan();
+		discoveryRunning = false;
+	}
+
+	@Override
+	protected void stopBackgroundDiscovery() {
+		discoveryRunning = false;
+	}
+
+	/**
+	 * Scans for Squeeze Servers
+	 */
+	private synchronized void scanServer() {
+		logger.debug("Scanning Server");
+		discoveryRunning = true;
+		Thread thread = new Thread("SqueezeBox Sendbroadcast") {
+			public void run() {
+				sendDiscoveryMessage();
+				discoveryRunning = false;
+				logger.trace("Done sending broadcast discovery messages.");
+			}
+		};
+
+		try {
+			// setup socket
+			startSocket();
+			thread.start();
+			receiveDiscoveryMessage();
+		} catch (Exception e) {
+			logger.debug("Problem scanning for Squeeze Server", e);
+		} finally {
+			// make sure we clean up.
+			try {
+				stopSocket();
+			} catch (Exception ignored) {
+
+			}
+		}
+	}
+
+	/**
+	 * Creates our UDP socket
+	 */
+	private void startSocket() throws Exception {
+
+		socket = new DatagramSocket();
```

Comment:
I probably do not know enough about UDP communication, but isn't it required to bind the socket to some UDP port to listen on?